Create RasterizationScale.md (#963)
MicrosoftEdge/WebView2Feedback: specs/RasterizationScale.md

we also know that there are consumers of WebView that want to explicitly set a custom DPI scale for WebView.

In this document we describe the updated API. We'd appreciate your feedback.

# Description
To allow the app to control the DPI scale of WebView, we are adding a `RasterizationScale` property to the CoreWebView2Controller interface.
The RasterizationScale property controls the DPI scaling for all UI in WebView.

In order to maintain compatibility with apps developed before this API existed,
WebView2 continues to detect monitor scale changes be default and will update the RasterizationScale property.
When the RasterizationScale property is updated, the `RasterizationScaleChanged` event is raised.
The app can tell WebView2 to stop updating the RasterizationScale property by changing
`ShouldDetectMonitorDpiScaleChanges` from the default value of true, to false.

# Examples
The following code snippets show how to use the RasterizationScale and ShouldDetectMonitorScaleChanges properties,
and now to listen to the RasterizationScaleChanged event.
## Win32 C++
```cpp
void ViewComponent::SetRasterizationScale(float additionalScale)
{

>     CHECK_FAILURE(m_controller->put_ShouldDetectMonitorScaleChanges(FALSE));

Reviewer: Do we need to set shoulddetectmonitorscalechanges to true in this c++ sample?

Author: This bit of code should be setting it to false because the app is taking control of setting the RasterizationScale property. The default value is TRUE. I don't expect apps to set it to true, unless for some reason they decided they no longer want to control the rasterization scale and want WebView to return to the default behavior.

Reviewer: Not sure what I meant by this comment.

    m_webviewAdditionalRasterizationScale = additionalScale;
    double rasterizationScale =
        additionalScale * m_appWindow->GetDpiScale() * m_appWindow->GetTextScale();
    CHECK_FAILURE(m_controller->put_RasterizationScale(rasterizationScale));
}

Callback<ICoreWebView2RasterizationScaleChangedEventHandler>(
    [this](ICoreWebView2Controller* sender, IUnknown* args) -> HRESULT {
        double rasterizationScale;
        CHECK_FAILURE(m_controller->get_RasterizationScale(&rasterizationScale));

        std::wstring message = L"WebView2APISample (RasterizationScale: " +
            std::to_wstring(int(rasterizationScale * 100)) + L"%)";
        SetWindowText(m_appWindow->GetMainWindow(), message.c_str());
        return S_OK;
    })
.Get(), &m_rasterizationScaleChangedToken));
```
## .Net
```c#
public void SetRasterizationScale(double rasterizaitonScale)
{
    CoreWebView2Controller.ShouldDetectMonitorScaleChanges = false;

>     CoreWebView2Controller.RasterizationScale = rasterizationScale;

Reviewer: Is this intended to be the C# equiv of the above? We don't need to factor in the various scales and things like we do in C++? If we don't need those then please add a comment or something explaining.

Author: Updated to match C++ sample.

}
```

# API Notes
See [API Details](#api-details) section below for API reference.

# API Details
## Win32 C++
```c#
interface ICoreWebView2Controller2 : ICoreWebView2Controller {
  /// The rasterization scale for the WebView. The rasterization scale is the
  /// combination of the monitor DPI scale and text scaling set by the user.
  /// This value should be updated when the DPI scale of the app's top level
  /// window changes (i.e. monitor DPI scale changes or window changes monitor)
  /// or when the text scale factor of the system changes.
  ///
  /// \snippet AppWindow.cpp DPIChanged
  ///
  /// \snippet AppWindow.cpp TextScaleChanged1
  ///
  /// \snippet AppWindow.cpp TextScaleChanged2
  ///
  /// Rasterization scale applies to the WebView content, as well as

>   /// popups, context menus, scroll bars, etc. Normal app scaling scenarios

Reviewer: The Microsoft style guide says to use 'and so on' instead of etc. Our documentation expert went through and replaced all our usage at some point iirc.

```suggestion
  /// popups, context menus, scroll bars, and so on. Normal app scaling scenarios
```

Author: Fixed and caught another place.

>   /// should use the ZoomFactor property or SetBoundsAndZoomFactor API.

Reviewer: ```suggestion
  /// should use the ZoomFactor property or SetBoundsAndZoomFactor API which
  /// only scale the rendered HTML content and not popups, context menus, scroll bars and so on.
```

Author: Updated.

  ///
  /// \snippet ViewComponent.cpp RasterizationScale
  [propget] HRESULT RasterizationScale([out, retval] double* scale);
  // Set the rasteriation scale property.
  [propput] HRESULT RasterizationScale([in] double scale);

  /// ShouldDetectMonitorScaleChanges property determines whether the WebView
  /// attempts to track monitor DPI scale changes. When true, the WebView will
  /// track monitor DPI scale changes, update the RasterizationScale property,
  /// and raises RasterizationScaleChanged event. When false, the WebView will
  /// not track monitor DPI scale changes, and the app must update the
  /// RasterizationScale property itself. RasterizationScaleChanged event will
  /// never raise when ShouldDetectMonitorScaleChanges is false.
  [propget] HRESULT ShouldDetectMonitorScaleChanges([out, retval] BOOL* value);
  /// Set the ShouldDetectMonitorScaleChanges property.

>   [propput] HRESULT ShouldDetectMonitorScaleChanges([in] BOOL value);

Reviewer: What happens when you modify RasterizationScale and ShouldDetectMonitorScaleChanges is true?

Author: The current implementation in experimental, is that it essentially does nothing. If the app sets the RasterizationScale while ShouldDetectMonitorScaleChanges is true, the WebView changes the RasterizationScale back to match the monitor scale and raises the RasterizationScaleChanged event.

We could change this behavior so that the RasterizationScale set by the app lasts until the monitor scales.  Or do you think the behavior is okay and just update the documentation to describe this behavior?

Reviewer: I don't think its a scenario we need to support and current behavior sounds fine. Just document it.

        /// Rasterization scale applies to the WebView content, as well as popups, context menus, scroll bars, etc. Normal app scaling scenarios should use the <see cref="CoreWebView2.ZoomFactor"/> property or <see cref="CoreWebView2.SetBoundsAndZoomFactor"/> method.
        /// </remarks>
        public double RasterizationScale { get; set;};

        /// <summary>
        /// Determines whether the WebView will detect monitor scale changes.
        /// </summary>
        /// <remarks>
        /// ShouldDetectMonitorScaleChanges property determines whether the WebView attempts to track monitor DPI scale schanges. When true, the WebView will track monitor DPI scale changes, update the <see cref="CoreWebView2.RasterizationScale"/> property, and fire <see cref="CoreWebView2.RasterizationScaleChanged"/> event. When false, the WebView will not track monitor DPI scale changes, and the app must update the <see cref="CoreWebView2.RasterizationScale"/> property itself. <see cref="CoreWebView2.RasterizationScaleChanged"/> event will never fire when ShouldDetectMonitorScaleChanges is false.
        /// </remarks>

>         public int ShouldDetectMonitorScaleChanges {get; set;};

Reviewer: Should be Boolean not int

Author: Looks like .Net is bool and WinRT is boolean. Since this came from .Net, I updated this to bool.

# Background
The WebView attempts to detects monitor scale changes and applies the monitor DPI scale to all UI produced by the WebView.
When the host app has a DPI awareness of PerMonitorV2, then all child HWNDs receive WM_DPICHANGED.
The WebView can check the current monitor DPI scale when it receives WM_DPICHANGED.
But if the host app is just PerMonitor aware (not v2), then only the top level window receives WM_DPICHANGED.
The WebView could try to use other metrics (like resize) to check the DPI, but resize is not guaranteed in that scenario.
From [GitHub feedback](https://github.com/MicrosoftEdge/WebView2Feedback/issues/65),
we also know that there are consumers of WebView that want to explicitly set a custom DPI scale for WebView.

In this document we describe the updated API. We'd appreciate your feedback.

# Description
To allow the app to control the DPI scale of WebView, we are adding a `RasterizationScale` property to the CoreWebView2Controller interface.
The RasterizationScale property controls the DPI scaling for all UI in WebView.

In order to maintain compatibility with apps developed before this API existed,

> WebView2 continues to detect monitor scale changes be default and will update the RasterizationScale property.

Reviewer: ```suggestion
WebView2 continues to detect monitor scale changes by default and will update the RasterizationScale property.
```

When the RasterizationScale property is updated, the `RasterizationScaleChanged` event is raised.
The app can tell WebView2 to stop updating the RasterizationScale property by changing
`ShouldDetectMonitorDpiScaleChanges` from the default value of true, to false.

# Examples
The following code snippets show how to use the RasterizationScale and ShouldDetectMonitorScaleChanges properties,
and now to listen to the RasterizationScaleChanged event.
## Win32 C++
```cpp
void ViewComponent::SetRasterizationScale(float additionalScale)
{
    CHECK_FAILURE(m_controller->put_ShouldDetectMonitorScaleChanges(FALSE));
    m_webviewAdditionalRasterizationScale = additionalScale;
    double rasterizationScale =

>         additionalScale * m_appWindow->GetDpiScale() * m_appWindow->GetTextScale();

Reviewer: Are these 3 scales common things we expect apps to have? If so, it would be good to describe the 3 values here in a comment.

Author: DPI scale and text scale come from the system and are controlled by the user in the Settings app. The IDL comments for RasterizationScale say that these two values are needed for RasterizationScale. The app has the option to add an additional scale if desired. I added a comment to the samples.
